crashsimilarity/downloader.py

```diff
@@ -0,0 +1,146 @@
+    _SUPER_SEARCH_URL = 'https://crash-stats.mozilla.com/api/SuperSearch'
+    _PROCESSED_CRASH_URL = 'https://crash-stats.mozilla.com/api/ProcessedCrash'
+
+    def __init__(self, cache=None):
+        super().__init__(cache)
+
+    def download_stack_traces_for_signature(self, signature, traces_num=100, period=timedelta(days=31)):
+        from_date = utils.utc_today() - period
+
+        key = ('traces_for_signature', signature, utils.utc_today(), from_date)
+        if self._cache and key in self._cache:
+            logging.debug('get data from cache')
+            return self._cache[key]
+
+        params = {
+            'signature': '=' + signature,
+            'date': ['>=' + str(from_date)],
+            '_facets': ['proto_signature'],
+            '_facets_size': traces_num,
+            '_results_number': 0
+        }
+        response = self.get_with_retries(self._SUPER_SEARCH_URL, params)
+        records = self._json_or_raise(response)['facets']['proto_signature']
+        traces = set([r['term'] for r in records])
+
+        if self._cache:
+            self._cache[key] = traces
+        return traces
+
+    def download_crash(self, uuid):
+        key = ('crash_for_uuid', uuid, utils.utc_today())
+        if self._cache and key in self._cache:
+            logging.debug('get data from cache')
+            return self._cache[key]
+
+        params = {'crash_id': uuid}
+        crash = self._json_or_raise(self.get_with_retries(self._PROCESSED_CRASH_URL, params))
+
+        if self._cache:
+            self._cache[key] = crash
+        return crash
+
+    def download_day_crashes(self, day, product='Firefox', offset=0, crashes_per_request=1000):
+        """While there can be ~100mb of data this function return generator"""
+        params = {
+            'product': product,
+            'date': ['>=' + str(day), '<' + str(day + timedelta(1))],
+            '_columns': ['uuid', 'signature', 'proto_signature'],
+            '_results_number': crashes_per_request,
+            '_results_offset': offset,
+            '_facets_size': 0,
+        }
+        logging.info('start downloading crashes for {}'.format(day))
+        uuids = set()
+        while True:
+            params['_results_offset'] = offset
+            response = self._json_or_raise(self.get_with_retries(self._SUPER_SEARCH_URL, params))
+            logging.info('offset: {} from: {}'.format(offset, response['total']))
+            crashes = response['hits']
+            offset += len(crashes)
+            for crash in crashes:
+                if crash['uuid'] not in uuids:
+                    uuids.add(crash['uuid'])
+                    yield crash
+            if len(crashes) < crashes_per_request:
+                break
+
+    @staticmethod
+    def download_and_save_crashes(days, product='Firefox', save_to_dir=utils.CRASHSIMILARITY_DATA_DIR):
+        if not os.path.exists(save_to_dir):
+            utils.create_dir(save_to_dir)
```

Comment:
Nit: you can use `utils.create_dir` without checking if the directory exists first (`utils.create_dir` is a no-op if the directory already exists).